src/hotspot/share/gc/shenandoah/shenandoahMmuTracker.cpp

```diff
@@ -70,49 +93,111 @@ double ShenandoahMmuTracker::process_time_seconds() {
+  double gcu = (gc_time - _most_recent_gc_time) / duration;
+  double mu = (mutator_time - _most_recent_mutator_time) / duration;
+  if (has_old_candidates) {
+    _doing_mixed_evacuations = true;
+  }
+  log_info(gc, ergo)("At end of %s: GC Utilization: %.1f%% for duration %.3fs (which is subsumed in next concurrent gc report)",
+                     old_marking_done? "last OLD marking increment": "OLD marking increment",
+                     _most_recent_gcu * 100, duration);
+}
+
+void ShenandoahMmuTracker::record_mixed(ShenandoahGeneration* generation, uint gcid, bool is_mixed_done) {
+  help_record_concurrent(generation, gcid, "Mixed Concurrent GC");
+}
+
+void ShenandoahMmuTracker::record_degenerated(ShenandoahGeneration* generation,
+                                              uint gcid, bool is_old_bootstrap, bool is_mixed_done) {
+  if ((gcid == _most_recent_gcid) && _most_recent_is_full) {
+    // Do nothing.  This is a redundant recording for the full gc that just completed.
+  } else if (is_old_bootstrap) {
+    help_record_concurrent(generation, gcid, "Degenerated Bootstrap Old GC");
+    if (!is_mixed_done) {
+      _doing_mixed_evacuations = true;
+    }
+  } else {
+    help_record_concurrent(generation, gcid, "Degenerated Young GC");
+  }
+}
+
+void ShenandoahMmuTracker::record_full(ShenandoahGeneration* generation, uint gcid) {
+  help_record_concurrent(generation, gcid, "Full GC");
+  _most_recent_is_full = true;
+  _doing_mixed_evacuations = false;
 }
 
 void ShenandoahMmuTracker::report() {
   // This is only called by the periodic thread.
-  double process_time_s = process_time_seconds();
-  double elapsed_process_time_s = process_time_s - _process_reference_time_s;
-  if (elapsed_process_time_s <= 0.01) {
-    // No cpu time for this interval?
-    return;
-  }
+  double current = os::elapsedTime();
+  double time_delta = current - _most_recent_periodic_time_stamp;
+  _most_recent_periodic_time_stamp = current;
+
+  double gc_time, mutator_time;
+  fetch_cpu_times(gc_time, mutator_time);
+
+  double gc_delta = gc_time - _most_recent_periodic_gc_time;
+  _most_recent_periodic_gc_time = gc_time;
 
-  _process_reference_time_s = process_time_s;
-  double collector_time_s = gc_thread_time_seconds();
-  double elapsed_collector_time_s = collector_time_s - _collector_reference_time_s;
-  _collector_reference_time_s = collector_time_s;
-  double minimum_mutator_utilization = ((elapsed_process_time_s - elapsed_collector_time_s) / elapsed_process_time_s) * 100;
-  _mmu_average.add(minimum_mutator_utilization);
-  log_info(gc)("Average MMU = %.3f", _mmu_average.davg());
+  double mutator_delta = mutator_time - _most_recent_periodic_mutator_time;
+  _most_recent_periodic_mutator_time = mutator_time;
+
+  double mu = mutator_delta / (_active_processors * time_delta);
+  double gcu = gc_delta / (_active_processors * time_delta);
+  log_info(gc)("Periodic Sample: Average GCU = %.3f%%, Average MU = %.3f%%", gcu * 100, mu * 100);
```

Comment:
These aren't really averages any more right? Just the utilization over the last period?